nomad/core_sched.go

```diff
@@ -706,9 +715,116 @@ func allocGCEligible(a *structs.Allocation, job *structs.Job, gcTime time.Time,
 	return timeDiff > interval.Nanoseconds()
 }
 
-// csiVolumeGC is used to garbage collect CSI volume publications
-func (c *CoreScheduler) csiVolumePublicationGC(eval *structs.Evaluation) error {
-	// TODO: implement me!
-	c.logger.Trace("garbage collecting unclaimed CSI volume publications")
+// csiVolumeClaimGC is used to garbage collect CSI volume claims
+func (c *CoreScheduler) csiVolumeClaimGC(eval *structs.Evaluation) error {
+	c.logger.Trace("garbage collecting unclaimed CSI volume claims")
+
+	// JobID smuggled in with the eval's own JobID
+	var jobID string
+	evalJobID := strings.Split(eval.JobID, ":")
+	if len(evalJobID) != 2 {
+		c.logger.Error("volume gc called without jobID")
+		return nil
+	}
+
+	jobID = evalJobID[1]
+	job, err := c.srv.State().JobByID(nil, eval.Namespace, jobID)
+	if err != nil || job == nil {
+		c.logger.Trace(
+			"cannot find job to perform volume claim GC. it may have been garbage collected",
+			"job", jobID)
+		return nil
+	}
+	c.volumeClaimReap([]*structs.Job{job}, eval.LeaderACL)
 	return nil
 }
+
+// volumeClaimReap contacts the leader and releases volume claims from
+// terminal allocs
+func (c *CoreScheduler) volumeClaimReap(jobs []*structs.Job, leaderACL string) error {
+	ws := memdb.NewWatchSet()
+	var result *multierror.Error
+
+	for _, job := range jobs {
+		c.logger.Trace("garbage collecting unclaimed CSI volume claims for job", "job", job.ID)
+		for _, taskGroup := range job.TaskGroups {
+			for _, tgVolume := range taskGroup.Volumes {
+				if tgVolume.Type != structs.VolumeTypeCSI {
+					continue // filter to just CSI volumes
+				}
+				volID := tgVolume.Source
+				vol, err := c.srv.State().CSIVolumeByID(ws, volID)
+				if err != nil {
+					result = multierror.Append(result, err)
+					continue
+				}
+				if vol == nil {
+					c.logger.Trace("cannot find volume to be GC'd. it may have been deregistered",
+						"volume", volID)
+					continue
+				}
+				vol, err = c.srv.State().CSIVolumeDenormalize(ws, vol)
+				if err != nil {
+					result = multierror.Append(result, err)
+					continue
+				}
+
+				gcAllocs := []*structs.Allocation{}
+				unclaimedNodes := map[string]struct{}{}
+
+				collectFunc := func(allocs map[string]*structs.Allocation) {
+					for _, alloc := range allocs {
+						// we call denormalize on the volume above to populate
+						// Allocation pointers. But the alloc might have been
+						// garbage collected concurrently, so if the alloc is
+						// still nil we can safely skip it.
+						if alloc == nil {
+							continue
+						}
+						if !alloc.Terminated() {
+							unclaimedNodes[alloc.NodeID] = struct{}{}
+							continue
+						}
+						gcAllocs = append(gcAllocs, alloc)
+						delete(unclaimedNodes, alloc.NodeID)
+					}
+				}
+
+				collectFunc(vol.WriteAllocs)
+				collectFunc(vol.ReadAllocs)
+
+				req := &structs.CSIVolumeClaimRequest{
+					VolumeID:   volID,
+					Allocation: nil, // unpublish never uses this field
```

Comment:
We do need the allocation for it's `ID` in order to release.

Reply:
We add that after the controller unpublish.

Reply:
Note from chat: https://github.com/hashicorp/nomad/pull/7167 changes this type to take `AllocationID` so if that merges first we'll need to fix this.

Reply:
Done. Rebased on `f-csi-volumes` and waiting for CI to pass.